docs/ingestion-beam/amplitude-job.md

```diff
@@ -0,0 +1,116 @@
+
+### Parse Payload
+
+Parse the message body as a `UTF-8` encoded JSON payload and drop some specific fields.
+
+### Parse User Agent
+
+Attempt to extract browser, browser version, and os from the `user_agent`
+attribute, drop any nulls, and remove `user_agent` from attributes.
+
+### Write Metadata Into the Payload
+
+Add a nested `metadata` field and several `normalized_*` attributes into the
+payload body.
+
+### Parse Amplitude Events
+
+This step reads an external CSV configuration file that contains a set of events that should be sent do Amplitude. The format of this file is as follows:
+
+```csv
+<namespace>,<doc type>,<event category>,<event name or '*'>
+```
+
+The `events` are getting parsed from the message payload and transformed into individual events that can be sent to the Amplitude API. Events that don't match the allow list are being filtered out.
+
+### Batching of Events
+
+Since the Amplitude API has [limitations](https://amplitude.com/docs/apis/analytics/batch-event-upload#considerations) around how many events can be sent in a single request and how many requests can be sent withing a second, this step creates batches of events.
+
+### Send request
+
+This step sends batched events to the Amplitude API.
+// todo: retry mechanism on failures which can happen when API limits are exceeded
+
+### Working with the Beam Job
+
+Options specific to this job are found in https://github.com/mozilla/gcp-ingestion/blob/main/ingestion-beam/src/main/java/com/mozilla/telemetry/amplitude/AmplitudePublisherOptions.java
+
+### Test Deployment
+
+This job can be deployed in a sandbox project for testing.
+
+There are a few required components to get a job running:
+
+- Upload a `.ndjson` file with example payload data to a GCS bucket
+- A event allowed list stored in GCS
+- Optionally, if reporting is enabled, a file with the Amplitude API key uploaded to GCS
+- The Beam pipeline running on Dataflow, reading from the input, and sending data to Amplitude
+
+Example script to start the Dataflow job from the ingestion-beam directory:
+
+```
+#!/bin/bash
+
+set -ux
+
+PROJECT="amplitude-dev"
+JOB_NAME="amplitude"
+path="$BUCKET/data/*.ndjson"
+
+mvn -X compile exec:java -Dexec.mainClass=com.mozilla.telemetry.AmplitudePublisher -Dexec.args="\
+    --runner=Dataflow \
+    --jobName=$JOB_NAME \
+    --project=$PROJECT  \
+    --geoCityDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-City/20241105/GeoIP2-City.mmdb \
+    --geoCityFilter=gs://moz-fx-data-prod-dataflow-templates/cities15000.txt \
+    --geoIspDatabase=gs://moz-fx-data-prod-geoip/GeoIP2-ISP/20241101/GeoIP2-ISP.mmdb \
+    --schemasLocation=gs://moz-fx-data-prod-dataflow/schemas/202411060452_e01d1666.tar.gz \
+    --inputType=file \
+    --input=$path \
+    --bqReadMethod=storageapi \
+    --outputType=bigquery \
+    --bqWriteMethod=file_loads \
+    --output=${PROJECT}:test.output_v1 \
+    --errorOutputType=bigquery \
+    --errorOutput=${PROJECT}:payload_bytes_error.structured \
```

Comment:
We don't want to write into `payload_bytes_error` in case we want to backfill/reprocess events

Reply:
This also conflicts with line above `--errorOutputType=bigquery \`.
BTW that's another option for dumping errors, easier to debug than Pub/Sub.